core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -98,55 +99,91 @@ object TransactionMarkerChannelManager {
   }
 
   def shutdown(): Unit = {
-    interBrokerSendThread.shutdown()
-    transactionMarkerChannel.clear()
+    txnMarkerSendThread.shutdown()
+    txnMarkerChannel.clear()
   }
 
-
-  def addTxnMarkerRequest(txnTopicPartition: Int, metadata: TransactionMetadata, coordinatorEpoch: Int, completionCallback: WriteTxnMarkerCallback): Unit = {
-    val metadataToWrite = metadata synchronized metadata.copy()
-
-    if (!transactionMarkerChannel.maybeAddPendingRequest(txnTopicPartition, metadata))
-      // TODO: Not sure this is the correct response here?
-      completionCallback(Errors.INVALID_TXN_STATE)
-    else {
-      val delayedTxnMarker = new DelayedTxnMarker(metadataToWrite, completionCallback)
-      txnMarkerPurgatory.tryCompleteElseWatch(delayedTxnMarker, Seq(metadata.pid))
-
-      val result = metadataToWrite.state match {
-        case PrepareCommit => TransactionResult.COMMIT
-        case PrepareAbort => TransactionResult.ABORT
-        case s => throw new IllegalStateException("Unexpected txn metadata state while writing markers: " + s)
+  def addTxnMarkersToSend(transactionalId: String,
+                          coordinatorEpoch: Int,
+                          txnResult: TransactionResult,
+                          txnMetadata: TransactionMetadata): Unit = {
+
+    def appendToLogCallback(error: Errors): Unit = {
+      error match {
+        case Errors.NONE =>
+          trace(s"Competed sending transaction markers for $transactionalId as $txnResult")
+
+          txnStateManager.getTransactionState(transactionalId) match {
+            case Some(epochAndMetadata) =>
+              epochAndMetadata synchronized {
+                if (epochAndMetadata.coordinatorEpoch == coordinatorEpoch) {
+                  debug(s"Updating $transactionalId's transaction state to $txnMetadata with coordinator epoch $coordinatorEpoch for $transactionalId succeeded")
+
+                  // try to append to the transaction log
+                  def retryAppendCallback(error: Errors): Unit =
+                    error match {
+                      case Errors.NONE =>
+                        trace(s"Completed transaction for $transactionalId with coordinator epoch $coordinatorEpoch, final state: state after commit: ${txnMetadata.state}")
+
+                      case Errors.NOT_COORDINATOR =>
+                        info(s"No longer the coordinator for transactionalId: $transactionalId while trying to append to transaction log, skip writing to transaction log")
+
+                      case Errors.COORDINATOR_NOT_AVAILABLE =>
+                        warn(s"Failed updating transaction state for $transactionalId when appending to transaction log due to ${error.exceptionName}. retrying")
+
+                        // retry appending
+                        txnStateManager.appendTransactionToLog(transactionalId, coordinatorEpoch, txnMetadata, retryAppendCallback)
+
+                      case errors: Errors =>
```

Comment:
When this happens usually ops people need to be involved to manually truncate the log before starting. But I think this code can also be improved a bit for a more graceful shutdown.

Do you have any ideas?